src/gluonts/mx/distribution/empirical_distribution.py

```diff
@@ -0,0 +1,258 @@
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+from typing import Optional, Tuple
+
+import numpy as np
+import mxnet as mx
+
+from gluonts.core.component import validated
+from gluonts.mx import Tensor
+
+from .distribution import Distribution, getF
+from .distribution_output import DistributionOutput, ArgProj
+
+
+class EmpiricalDistribution(Distribution):
+    r"""
+    A class representing empirical distribution.
+    By default, the target is assumed to be vector-valued, i.e., `event_shape` is 1. However, note that each dimension
+    is assumed to be independent when computing variance and CRPS loss.
+
+    Also, for computing CDF and quantiles, it is assumede that samples are distinct along the samples dimension, which
+    should almost always be the case if samples are drawn from continuous distributions.
+
+    Parameters
+    ----------
+    samples
+        Tensor containing samples, of shape `(num_samples, *batch_shape, *event_shape)`.
+    event_dim
+        Number of event dimensions, i.e., length of the `event_shape` tuple.
+        This is `0` for distributions over scalars, `1` over vectors, `2` over matrices, and so on.
+    """
+
+    @validated()
+    def __init__(self, samples: Tensor, event_dim: int = 1) -> None:
+        self.samples = samples
+        self.sorted_samples = self.F.sort(self.samples, axis=0)
+        self.sorted_ix = self.F.argsort(self.samples, axis=0)
+
+        self._event_dim = event_dim
+        assert (
+            len(self.samples.shape) >= 1 + event_dim
+        ), "Shape of samples do not match with the value given for `event_dim`!"
+
+    @property
+    def F(self):
+        return getF(self.samples)
+
+    @property
+    def batch_shape(self) -> Tuple:
+        if self.event_dim == 0:
+            return self.samples.shape[1:]
+        else:
+            return self.samples.shape[1 : -self.event_dim]
+
+    @property
+    def event_shape(self) -> Tuple:
+        if self.event_dim == 0:
+            return ()
+        else:
+            return self.samples.shape[-self.event_dim :]
+
+    @property
+    def event_dim(self) -> int:
+        return self._event_dim
+
+    @property
+    def mean(self) -> Tensor:
+        return self.F.mean(self.samples, axis=0)
+
+    @property
+    def stddev(self) -> Tensor:
+        F = self.F
+        return F.sqrt(F.mean(F.square(self.mean - self.samples), axis=0))
+
+    def cdf(self, x: Tensor):
+        # Note: computes CDF on each dimension of the target independently.
+        self.CDF_sorted = self.F.linspace(
+            start=1 / self.samples.shape[0],
+            stop=1,
+            endpoint=True,
+            num=self.samples.shape[0],
+        )
+
+        # Replace this when mxnet is updated to a version where `searchsorted` is available!
```

Comment:
:-( wouldn’t it be better to rely on numpy for this one?

Reply:
I wanted to use numpy first, but looks like it only supports 1D arrays.